src/config.c

```diff
@@ -1552,60 +1552,47 @@ void rewriteConfigRemoveOrphaned(struct rewriteConfigState *state) {
- *
- * This way we are sure the file is left in a consistent state even if the
- * process is stopped between any of the four operations.
+/* This function replaces the old configuration file with the new content
+ * in an atomic manner.
  *
  * The function returns 0 on success, otherwise -1 is returned and errno
- * set accordingly. */
+ * is set accordingly. */
 int rewriteConfigOverwriteFile(char *configfile, sds content) {
-    int retval = 0;
-    int fd = open(configfile,O_RDWR|O_CREAT,0644);
-    int content_size = sdslen(content), padding = 0;
-    struct stat sb;
-    sds content_padded;
-
-    /* 1) Open the old file (or create a new one if it does not
-     *    exist), get the size. */
-    if (fd == -1) return -1; /* errno set by open(). */
-    if (fstat(fd,&sb) == -1) {
-        close(fd);
-        return -1; /* errno set by fstat(). */
-    }
-
-    /* 2) Pad the content at least match the old file size. */
-    content_padded = sdsdup(content);
-    if (content_size < sb.st_size) {
-        /* If the old file was bigger, pad the content with
-         * a newline plus as many "#" chars as required. */
-        padding = sb.st_size - content_size;
-        content_padded = sdsgrowzero(content_padded,sb.st_size);
-        content_padded[content_size] = '\n';
-        memset(content_padded+content_size+1,'#',padding-1);
-    }
-
-    /* 3) Write the new content using a single write(2). */
-    if (write(fd,content_padded,strlen(content_padded)) == -1) {
-        retval = -1;
-        goto cleanup;
-    }
-
-    /* 4) Truncate the file to the right length if we used padding. */
-    if (padding) {
-        if (ftruncate(fd,content_size) == -1) {
-            /* Non critical error... */
-        }
-    }
+    int fd = -1;
+    int retval = -1;
+    char tmp_config_path[PATH_MAX];
+    const char *tmp_suffix = ".XXXXXX";
+    size_t config_file_len = strlen(configfile);
+
+    if (config_file_len > PATH_MAX - strlen(tmp_suffix) - 1) {
+        serverLog(LL_WARNING, "Config file full path is too long");
+        errno = ENAMETOOLONG;
+        return retval;
+    }
+    strncpy(tmp_config_path, configfile, config_file_len);
+    strncpy(tmp_config_path + config_file_len, tmp_suffix, strlen(tmp_suffix)+1);
+
+#ifdef _GNU_SOURCE
+    fd = mkostemp(tmp_config_path, O_CLOEXEC);
+#else
+    /* There's a theoretical chance here to leak the FD if a module thread forks & execv in the middle */
+    fd = mkstemp(tmp_config_path);
+#endif
 
-cleanup:
-    sdsfree(content_padded);
-    close(fd);
+    if (fd == -1)
+        serverLog(LL_WARNING, "Could not create tmp config file (%s)", strerror(errno));
+    else if (write(fd,content,sdslen(content)) == -1)
+        serverLog(LL_WARNING, "Could not write to config file (%s)", strerror(errno));
+    else if (rename(tmp_config_path, configfile) == -1)
+        serverLog(LL_WARNING, "Could not rename tmp config file (%s)", strerror(errno));
+    else if (fchmod(fd, 0644) == -1)
```

Comment:
I think it might make sense to `fchmod()` earlier, so once `rename()` is successful we're all done. In this order we may end up with the wrong permissions.

Reply:
Good idea. Done.